### ISSUE TITLE
CA1065 warning in implicit operator

### ISSUE THREAD

**Issue by lonix1** [dotnet/roslyn-analyzers#5825]
### Analyzer

**Diagnostic ID**: [CA1065](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1065): `Do not raise exceptions in unexpected locations`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

_OR_

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 (Latest)

### Describe the bug

````cs
public class Word
{
	public Word(string s) => Value = s;
	public readonly string Value;
	public static implicit operator string (Word instance) => instance.Value;  // CA1062: Validate arguments of public methods
}
````

### Steps To Reproduce

The `instance` argument can be null in this case:
````cs
Word word = null;
string s = word;    // throws NullReferenceException
````

So let's fix it:
````cs
public static implicit operator string (Word instance) {
  _ = instance ?? throw new ArgumentNullException(nameof(instance));     // CA1065
  return instance.Value;
}
````
Now it complains with a `CA1065`:

> op_Implicit creates an exception of type ArgumentNullException. Exceptions should not be raised in this type of method. If this exception instance might be raised, change this method's logic so it no longer raises an exception.

### Expected behavior

Unsure - but I expect `CA1065` should be allowed?

### Actual behavior
as above

## Additional context
no

**Comment by lonix1** [comment]
Interesting point!

But it just doesn't feel exactly right. I think the client code, especially in a nullable reference environment, and especially in a DDD codebase that expects "always valid" instances, that would be surprising. On the other hand, you could say it's the client code that sent in the null, so its the client's fault.

This is definitely an edge case, there is no obvious "best" approach.

**Comment by Evangelink** [comment]
> there is no obvious "best" approach.

Agreed.

> especially in a nullable reference environment

This one is tricky because it doesn't prevent a null to be passed, it just somehow assume that if all your code is setup correctly you should not have null. But I tend to agree that in nullable context `CA1062` should probably not trigger. cc @mavasani 

If you are using recent enough version of C#, you could use parameter null checking `!!` to workaround the issue as `CA1065` doesn't recognize this syntax.

**Comment by Youssef1313** [comment]
@Evangelink The API in example is public, so really nothing prevents a null. I think CA1062 should still be reported in nullable-enabled context. This is similar to https://github.com/dotnet/roslyn-analyzers/issues/5350
